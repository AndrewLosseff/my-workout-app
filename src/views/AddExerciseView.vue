<script setup lang="ts">
import { ref, computed, onMounted } from 'vue'
import { useRouter, useRoute } from 'vue-router'
import { useWorkoutStore } from '../stores/workout'
import { useExerciseStore } from '../stores/exercise'

const router = useRouter()
const route = useRoute()
const workoutStore = useWorkoutStore()
const exerciseStore = useExerciseStore()
const form = ref(null)

const workoutId = computed(() => route.params.id as string)

onMounted(() => {
  workoutStore.loadFromLocalStorage()
})

const workout = computed(() => workoutStore.getWorkoutById(workoutId.value))

const exercise = ref({
  workoutId: workoutId.value,
  name: '',
  sets: 3,
  reps: 10,
  weight: 0,
})

const saveExercise = () => {
  // @ts-ignore - form validation
  if (!form.value.validate()) return

  exerciseStore.addExercise(exercise.value)
  router.push(`/workout/${workoutId.value}`)
}

const goBack = () => {
  router.push(`/workout/${workoutId.value}`)
}
</script>

<template>
  <div v-if="workout">
    <div class="d-flex align-center mb-4">
      <v-btn icon @click="goBack" class="mr-2">
        <v-icon>mdi-arrow-left</v-icon>
      </v-btn>
      <h1 class="text-h4">Add Exercise to {{ workout.name }}</h1>
    </div>

    <v-form @submit.prevent="saveExercise" ref="form">
      <v-card class="pa-4">
        <v-text-field
          v-model="exercise.name"
          label="Exercise Name"
          :rules="[(v) => !!v || 'Name is required']"
          required
        ></v-text-field>

        <v-row>
          <v-col cols="12" sm="4">
            <v-text-field
              v-model.number="exercise.sets"
              label="Sets"
              type="number"
              min="1"
              :rules="[
                (v) => !!v || 'Sets is required',
                (v) => v > 0 || 'Sets must be greater than 0',
              ]"
              required
            ></v-text-field>
          </v-col>

          <v-col cols="12" sm="4">
            <v-text-field
              v-model.number="exercise.reps"
              label="Reps"
              type="number"
              min="1"
              :rules="[
                (v) => !!v || 'Reps is required',
                (v) => v > 0 || 'Reps must be greater than 0',
              ]"
              required
            ></v-text-field>
          </v-col>

          <v-col cols="12" sm="4">
            <v-text-field
              v-model.number="exercise.weight"
              label="Weight (kg)"
              type="number"
              min="0"
              :rules="[
                (v) => !!v || 'Weight is required',
                (v) => v >= 0 || 'Weight must be 0 or greater',
              ]"
              required
            ></v-text-field>
          </v-col>
        </v-row>

        <v-card-actions>
          <v-spacer></v-spacer>
          <v-btn color="primary" type="submit">Save Exercise</v-btn>
          <v-btn color="error" @click="goBack">Cancel</v-btn>
        </v-card-actions>
      </v-card>
    </v-form>
  </div>
  <div v-else>
    <v-alert type="error">Workout not found</v-alert>
    <v-btn color="primary" to="/">Back to Dashboard</v-btn>
  </div>
</template>
