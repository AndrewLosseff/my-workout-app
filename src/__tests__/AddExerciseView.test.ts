import { describe, it, expect, beforeEach, vi } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import AddExerciseView from '@/views/AddExerciseView.vue'

vi.mock('vue-router', () => ({
  useRouter: () => ({
    push: vi.fn()
  }),
  useRoute: () => ({
    params: {
      id: '123'
    }
  })
}))

vi.mock('@/stores/workout', () => ({
  useWorkoutStore: () => ({
    loadFromLocalStorage: vi.fn(),
    getWorkoutById: vi.fn().mockReturnValue({
      id: '123',
      name: 'Test Workout',
      exercises: []
    })
  })
}))

vi.mock('@/stores/exercise', () => ({
  useExerciseStore: () => ({
    addExercise: vi.fn()
  })
}))

describe('AddExerciseView', () => {
  let wrapper: VueWrapper

  beforeEach(() => {
    setActivePinia(createPinia())

    wrapper = mount(AddExerciseView, {
      global: {
        stubs: {
          'v-btn': true,
          'v-icon': true,
          'v-form': true,
          'v-card': true,
          'v-text-field': true,
          'v-row': true,
          'v-col': true,
          'v-card-actions': true,
          'v-spacer': true,
          'v-alert': true
        }
      }
    })
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('displays the workout name in the title', () => {
    expect(wrapper.find('h1').text()).toContain('Add Exercise to Test Workout')
  })

  it('initializes the form with default values', () => {
    expect(wrapper.vm.exercise.name).toBe('')
    expect(wrapper.vm.exercise.sets).toBe(3)
    expect(wrapper.vm.exercise.reps).toBe(10)
    expect(wrapper.vm.exercise.weight).toBe(0)
  })

  it('navigates back when back button is clicked', async () => {
    const backButton = wrapper.find('v-btn-stub[icon]')
    await backButton.trigger('click')

    expect(wrapper.vm.router.push).toHaveBeenCalledWith('/workout/123')
  })

  it('adds exercise and navigates on successful form submission', async () => {
    wrapper.vm.form = { validate: () => true }

    wrapper.vm.exercise.name = 'Bench Press'

    await wrapper.find('v-form-stub').trigger('submit')

    expect(wrapper.vm.exerciseStore.addExercise).toHaveBeenCalledWith({
      workoutId: '123',
      name: 'Bench Press',
      sets: 3,
      reps: 10,
      weight: 0
    })

    expect(wrapper.vm.router.push).toHaveBeenCalledWith('/workout/123')
  })

})
