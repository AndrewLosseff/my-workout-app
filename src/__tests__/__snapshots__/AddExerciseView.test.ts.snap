// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddExerciseView > renders correctly 1`] = `
"<div>
  <div class="d-flex align-center mb-4">
    <v-btn-stub icon="" class="mr-2"></v-btn-stub>
    <h1 class="text-h4">Add Exercise to Test Workout</h1>
  </div>
  <v-form-stub></v-form-stub>
</div>"
`;
